### PR TITLE
Implement CASE processing of IPK

### LINKED ISSUE #15583: CASESession does not make use of real IPK
#### Problem
The Identity Protection Key (IPK) is a critical component of CASE security in Matter. SDK currently uses a fixed value of all zeros, rather than using the IPK logic from the spec

 #### Proposed Solution
Implement IPK processing by:

- Using current IPK for fabric from Group Data Provider when generating DestinationID in Sigma 1 for transmit
  - Recording which of the IPKs it was since it will be needed for subsequent key derivation steps
- Properly matching DestinationID using the installed IPKs in processing Sigma1.
  - Recording which IPK was used on receipt when processing Sigma1, since it is required in steps after Sigma1 in several key derivations

### PR DESCRIPTION
**Pull request by tcarmelveilleux**
#### Problem
- IPK processing used placeholder IPK at innermost levels of
  CASE and never used the values set in GroupDataProvider as
  set by Commissioner with AddNOC or KeySetWrite cluster
  commands.
- IPK Processing requires significant state keeping by
  controllers and commissioners to work, and none of the
  plumbing existed

Fixes #15583

#### Change overview
- Fixed NOC Cluster setting of IPK in AddNOC that was using
  the wrong compressed fabric ID.
- Properly use GroupDataProvider to get IPK for controllers
  sending Sigma1
- Properly iterate through all IPK on receiving Sigma1, from
  GroupDataProvider
- Add plumbing to properly initialize GroupDataProvider at
  CHIPDeviceControllerFactory and Server
- Added a central point for a default IPK, that is at the very
  outermost level, rather than innermost level. A follow-up
  PR will allow reconfiguration of it for CHIP-tool. Code paths
  will properly use the OperationalCredentialsDelegate's IPK
  value passed in Callback, if you it's provided. Controllers and
  commissioners can setup their GroupDataProvider context to
  properly use the right group keys even without the
  follow-up to allow non-default IPK in chip-tool.
- Cleaned-up the loose ends around all the injection points

#### Testing
- Updated all necessary Unit tests
- Added IPK and Destination ID unit test cases from spec
- All cert tests still pass
- All unit tests still pass

### CODE REVIEW

File: src/controller/python/ChipDeviceController-ScriptBinding.cpp [hunk 1]
```diff
@@ -216,6 +218,14 @@ ChipError::StorageType pychip_DeviceController_StackInit()
...
     FactoryInitParams factoryParams;
     factoryParams.fabricIndependentStorage = sStorageAdapter;
+
+    // TODO: Provide a means to Python-configure the group data provider to update local IPK
+    sGroupDataProvider.SetStorageDelegate(&sStorageAdapter);
+    ReturnErrorOnFailure(sGroupDataProvider.Init().AsInteger());
+
+    // TODO: Set IPK from src/lib/support/TestGroupData.h:DefaultIpkValue::GetDefaultIpk()
+
+    factoryParams.groupDataProvider        = sGroupDataProvider;
```
Comment (tcarmelveilleux) [inline src/controller/python/ChipDeviceController-ScriptBinding.cpp:228]: @erjiaqing Can you please see if you can properly init this file and `src/controller/python/chip/internal/CommissionerImpl.cpp` to have the IPK (see https://github.com/project-chip/connectedhomeip/pull/16737/files#diff-7d33bc2a1145c23e5f424c29d4377e9dd9845f4db0b0ebdfa34286647b39f831R126)

File: examples/platform/linux/AppMain.cpp [hunk 2]
```diff
@@ -427,6 +432,20 @@ CHIP_ERROR InitCommissioner()
...
+
+    uint8_t compressedFabricId[sizeof(uint64_t)] = { 0 };
+    MutableByteSpan compressedFabricIdSpan(compressedFabricId);
+    ReturnErrorOnFailure(fabricInfo->GetCompressedId(compressedFabricIdSpan));
+    ChipLogProgress(Support, "Setting up group data with Compressed Fabric ID:");
+    ChipLogByteSpan(Support, compressedFabricIdSpan);
+
+    ByteSpan defaultIpk = chip::GroupTesting::DefaultIpkValue::GetDefaultIpk();
+    ReturnLogErrorOnFailure(chip::Credentials::SetSingleIpkEpochKey(&gGroupDataProvider, fabricInfo->GetFabricIndex(), defaultIpk,
+                                                                    compressedFabricIdSpan));
```
Comment (msandstedt) [inline examples/platform/linux/AppMain.cpp:447]: For this self-commissioned variant of the the Linux app, we should have a TODO for the credential issuer to initialize the gGroupDataProvider.

File: src/app/clusters/ethernet-network-diagnostics-server/ethernet-network-diagnostics-server.cpp [hunk 3]
```diff
@@ -76,6 +76,8 @@ CHIP_ERROR EthernetDiagosticsAttrAccess::ReadPHYRate(AttributeValueEncoder & aEn
     Attributes::PHYRate::TypeInfo::Type pHYRate;
     PHYRateType value = EmberAfPHYRateType::EMBER_ZCL_PHY_RATE_TYPE_10_M;
 
+    pHYRate.SetNull();
```
Comment (msandstedt) [inline src/app/clusters/ethernet-network-diagnostics-server/ethernet-network-diagnostics-server.cpp:79]: Is this change intentional?

File: src/credentials/tests/TestGroupDataProvider.cpp [hunk 4]
```diff
@@ -797,17 +816,70 @@ void TestKeySets(nlTestSuite * apSuite, void * apContext)
...
+                                             0xf2, 0x3b, 0xa4, 0xbd, 0x6b, 0x9d, 0xd9, 0x60 };
+
+    NL_TEST_ASSERT(apSuite, 0 == ipkOperationalKeySet.keyset_id);
+    NL_TEST_ASSERT(apSuite, 1 == ipkOperationalKeySet.num_keys_used);
+    NL_TEST_ASSERT(apSuite, SecurityPolicy::kTrustFirst == ipkOperationalKeySet.policy);
+    NL_TEST_ASSERT(apSuite, 1234 == ipkOperationalKeySet.epoch_keys[0].start_time);
+    NL_TEST_ASSERT(apSuite,
+                   0 == memcmp(ipkOperationalKeySet.epoch_keys[0].key, kExpectedIpkFromSpec, sizeof(kExpectedIpkFromSpec)));
+
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == provider->RemoveKeySet(kFabric1, kKeysetId0));
```
Comment (msandstedt) [inline src/credentials/tests/TestGroupDataProvider.cpp:857]:
```suggestion

    // Remove IPK, verify removal
    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == provider->RemoveKeySet(kFabric1, kKeysetId0));
```

File: src/credentials/tests/TestGroupDataProvider.cpp [hunk 5]
```diff
@@ -797,17 +816,70 @@ void TestKeySets(nlTestSuite * apSuite, void * apContext)
...
+    NL_TEST_ASSERT(apSuite, CHIP_ERROR_NOT_FOUND == provider->GetIpkKeySet(kFabric1, ipkOperationalKeySet));
+
+    NL_TEST_ASSERT(
```
Comment (msandstedt) [inline src/credentials/tests/TestGroupDataProvider.cpp:860]:
```suggestion

    // Set a single IPK with the SetSingleIpkEpochKey helper, validate key derivation follows spec
    NL_TEST_ASSERT(
```

File: src/app/clusters/ethernet-network-diagnostics-server/ethernet-network-diagnostics-server.cpp (hunk 3 above)
Comment (tcarmelveilleux) [reply src/app/clusters/ethernet-network-diagnostics-server/ethernet-network-diagnostics-server.cpp:79]: Yes, CI was failing esp32 due to non initialized

Comment (bzbarsky-apple) [reply src/app/clusters/ethernet-network-diagnostics-server/ethernet-network-diagnostics-server.cpp:79]: @tcarmelveilleux Which CI was failing?  I thought we had made that warning not fail in all the cases that matter....

In particular, `pHYRate`'s constructor already did the equivalent of `SetNull`.....

File: examples/chip-tool/commands/common/CHIPCommand.cpp [hunk 6]
```diff
@@ -65,8 +65,20 @@ CHIP_ERROR CHIPCommand::Run()
...
+
+    // Init group data provider that will be used for all group keys and IPKs for the
+    // chip-tool-configured fabrics. This is OK to do once since the fabric tables
+    // and the DeviceControllerFactory all "share" in the same underlying data.
+    // Different commissioner implementations may want to use alternate implementations
+    // of GroupDataProvider for injection through factoryInitParams.
+    mGroupDataProvider.SetStorageDelegate(&mDefaultStorage);
+    ReturnLogErrorOnFailure(mGroupDataProvider.Init());
+    chip::Credentials::SetGroupDataProvider(&mGroupDataProvider);
+    factoryInitParams.groupDataProvider = &mGroupDataProvider;
```
Comment (rcasallas-silabs) [inline examples/chip-tool/commands/common/CHIPCommand.cpp:79]: Since the data provider is a singleton accessible with `chip::Credentials::GetGroupDataProvider()`, this parameter seems redundant. Having two ways to access the data provider is a path to inconsistencies. I actually prefer proper injection, keeping the parameter, and removing the global `GetGroupDataProvider()`.

File: src/app/clusters/ethernet-network-diagnostics-server/ethernet-network-diagnostics-server.cpp (hunk 3 above)
Comment (tcarmelveilleux) [reply src/app/clusters/ethernet-network-diagnostics-server/ethernet-network-diagnostics-server.cpp:79]: It's because `Attributes::PHYRate::TypeInfo::Type pHYRate;` is never assigned a value. Somehow the compiler used for ESP32 tests blew-up on that. Let me take it out and find the run again.

File: examples/chip-tool/commands/common/CHIPCommand.cpp (hunk 6 above)
Comment (tcarmelveilleux) [reply examples/chip-tool/commands/common/CHIPCommand.cpp:79]: Yes, I would like to remove GetGroupDataProvider, but I want to do it as a follow-up, since there is more surgery needed. It will definitely be possible after this. What's missing is having the `SessionManager` get a reference so that it can use the keys properly, without accessing the singleton getter.

File: examples/platform/linux/AppMain.cpp [hunk 7]
```diff
@@ -427,6 +432,20 @@ CHIP_ERROR InitCommissioner()
...
+    Credentials::FabricInfo * fabricInfo = gCommissioner.GetFabricInfo();
+    VerifyOrReturnError(fabricInfo != nullptr);
+
+    uint8_t compressedFabricId[sizeof(uint64_t)] = { 0 };
+    MutableByteSpan compressedFabricIdSpan(compressedFabricId);
+    ReturnErrorOnFailure(fabricInfo->GetCompressedId(compressedFabricIdSpan));
+    ChipLogProgress(Support, "Setting up group data with Compressed Fabric ID:");
+    ChipLogByteSpan(Support, compressedFabricIdSpan);
+
+    ByteSpan defaultIpk = chip::GroupTesting::DefaultIpkValue::GetDefaultIpk();
```
Comment (tcarmelveilleux) [inline examples/platform/linux/AppMain.cpp:445]:
```suggestion
    // TODO: Once ExampleOperationalCredentialsIssuer has support, set default IPK on it as well so
    // that commissioned devices get the IPK set from real values rather than "test-only" internal hookups.
    ByteSpan defaultIpk = chip::GroupTesting::DefaultIpkValue::GetDefaultIpk();
```

File: src/app/server/Server.h [hunk 8]
```diff
@@ -203,22 +205,23 @@ class Server
...
-            mSessionManager = sessionManager;
+            mServer = server;
             return CHIP_NO_ERROR;
         };
 
         void OnFabricDeletedFromStorage(CompressedFabricId compressedId, FabricIndex fabricIndex) override
         {
             (void) compressedId;
-            if (mSessionManager != nullptr)
+            auto * sessionManager = &(mServer->GetSecureSessionManager());
```
Comment (rcasallas-silabs) [inline src/app/server/Server.h:219]: The reference here ensures that the pointer is not null. Why not use the reference directly? Should `GetSecureSessionManager()` return an pointer instead?

File: src/credentials/FabricTable.h [hunk 9]
```diff
@@ -110,7 +110,7 @@ class DLL_EXPORT FabricInfo
 
     void SetVendorId(uint16_t vendorId) { mVendorId = vendorId; }
 
-    Crypto::P256Keypair * GetOperationalKey()
+    Crypto::P256Keypair * GetOperationalKey() const
```
[the remark this answers is not in the document]
Comment (msandstedt) [reply src/credentials/FabricTable.h:113]: Ah, missed that `mOperationalKey` is marked mutable.

**Review by msandstedt — APPROVED** [review]

File: src/app/server/Server.h (hunk 8 above)
Comment (tcarmelveilleux) [reply src/app/server/Server.h:219]: `mServer` cannot be null based on Init()

File: examples/platform/linux/AppMain.cpp [hunk 10]
```diff
@@ -427,6 +433,23 @@ CHIP_ERROR InitCommissioner()
     auto & factory = Controller::DeviceControllerFactory::GetInstance();
     ReturnErrorOnFailure(factory.Init(factoryParams));
     ReturnErrorOnFailure(factory.SetupCommissioner(params, gCommissioner));
+
+    chip::FabricInfo * fabricInfo = gCommissioner.GetFabricInfo();
+    VerifyOrReturnError(fabricInfo != nullptr, CHIP_ERROR_INTERNAL);
+
+    uint8_t compressedFabricId[sizeof(uint64_t)] = { 0 };
```
Comment (mrjerryjohns) [inline examples/platform/linux/AppMain.cpp:440]: It seems like all the impls have pretty much the same bits of logic copied here. Perhaps all of this bit should be shifted into the `DeviceControllerFactory::Init`?

Comment (tcarmelveilleux) [reply examples/platform/linux/AppMain.cpp:440]: This is only available after CommissionerInit, which is done much later than DeviceControllerFactory::Init.

We can restructure the repeated init in a follow-up PR

**Review by rcasallas-silabs — APPROVED** [review]

File: src/protocols/secure_channel/CASESession.cpp [hunk 11]
```diff
@@ -342,13 +373,22 @@ CHIP_ERROR CASESession::SendSigma1()
...
-        ReturnErrorOnFailure(
-            mFabricInfo->GenerateDestinationID(ByteSpan(mIPK), ByteSpan(mInitiatorRandom), GetPeerNodeId(), destinationIdSpan));
+
+        // Obtain originator IPK matching the fabric where we are trying to open a session. mIPK
+        // will be properly set thereafter.
+        ReturnErrorOnFailure(RecoverInitiatorIpk());
+
+        FabricId fabricId = mFabricInfo->GetFabricId();
+        uint8_t rootPubKeyBuf[Crypto::kP256_Point_Length];
+        Credentials::P256PublicKeySpan rootPubKeySpan(&rootPubKeyBuf[0]);
```
Comment (bzbarsky-apple) [inline src/protocols/secure_channel/CASESession.cpp:386]:
```suggestion
        Credentials::P256PublicKeySpan rootPubKeySpan(rootPubKeyBuf);
```
which will helpfully ensure at compile time the buffer is big enough and whatnot...

File: src/protocols/secure_channel/CASESession.cpp [hunk 12]
```diff
@@ -402,6 +442,58 @@ CHIP_ERROR CASESession::HandleSigma1_and_SendSigma2(System::PacketBufferHandle &
...
+    VerifyOrReturnError(mFabricsTable != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    bool found = false;
+    for (const FabricInfo & fabricInfo : *mFabricsTable)
+    {
+        // Basic data for candidate fabric, used to compute candidate destination identifiers
+        FabricId fabricId = fabricInfo.GetFabricId();
+        NodeId nodeId     = fabricInfo.GetNodeId();
+        uint8_t rootPubKeyBuf[Crypto::kP256_Point_Length];
+        Credentials::P256PublicKeySpan rootPubKeySpan(&rootPubKeyBuf[0]);
```
Comment (bzbarsky-apple) [inline src/protocols/secure_channel/CASESession.cpp:456]:
```suggestion
        Credentials::P256PublicKeySpan rootPubKeySpan(rootPubKeyBuf);
```

File: src/protocols/secure_channel/CASESession.h [hunk 13]
```diff
@@ -110,6 +107,17 @@ class DLL_EXPORT CASESession : public Messaging::ExchangeDelegate, public Pairin
                      Messaging::ExchangeContext * exchangeCtxt, SessionEstablishmentDelegate * delegate,
                      Optional<ReliableMessageProtocolConfig> mrpConfig = Optional<ReliableMessageProtocolConfig>::Missing());
 
+    /**
+     * @brief Set the Group Data Provider which will be used to look-up IPKs
```
Comment (bzbarsky-apple) [inline src/protocols/secure_channel/CASESession.h:111]:
```suggestion
     * @brief Set the Group Data Provider which will be used to look up IPKs
```

File: src/protocols/secure_channel/CASESession.h [hunk 14]
```diff
@@ -184,6 +192,12 @@ class DLL_EXPORT CASESession : public Messaging::ExchangeDelegate, public Pairin
 
     CHIP_ERROR Init(uint16_t mySessionId, SessionEstablishmentDelegate * delegate);
 
+    // On success, sets mIpk to the correct value for outgoing Sigma1 based on internal state
+    CHIP_ERROR RecoverInitiatorIpk();
+    // On success, sets locally maching mFabricInfo in internal state to the entry matched by
+    // destinationId/initiatorRandom from processing of Sigma1
```
Comment (bzbarsky-apple) [inline src/protocols/secure_channel/CASESession.h:198]:
```suggestion
    // destinationId/initiatorRandom from processing of Sigma1, and sets mIpk to the right IPK.
```

**Review by bzbarsky-apple — APPROVED** [review]

File: src/credentials/FabricTable.cpp [hunk 15]
```diff
@@ -282,12 +282,7 @@ CHIP_ERROR FabricInfo::GeneratePeerId(FabricId fabricId, NodeId nodeId, PeerId *
         rootPubkey = rootPubkeySpan;
     }
 
-    ChipLogDetail(Inet, "Generating compressed fabric ID using uncompressed fabric ID 0x" ChipLogFormatX64 " and root pubkey",
-                  ChipLogValueX64(fabricId));
-    ChipLogByteSpan(Inet, ByteSpan(rootPubkey.ConstBytes(), rootPubkey.Length()));
     ReturnErrorOnFailure(GenerateCompressedFabricId(rootPubkey, fabricId, compressedFabricIdSpan));
-    ChipLogDetail(Inet, "Generated compressed fabric ID");
-    ChipLogByteSpan(Inet, compressedFabricIdSpan);
```
[the remark this answers is not in the document]
Comment (tcarmelveilleux) [reply src/credentials/FabricTable.cpp:290]: Because these logs are very very chatty and serve no useful purpose, and they were filling-up the areas of logs where I needed to understand what's happening.

File: src/app/OperationalDeviceProxy.h [hunk 16]
```diff
@@ -91,10 +94,15 @@ class DLL_EXPORT OperationalDeviceProxy : public DeviceProxy,
     ~OperationalDeviceProxy() override;
     OperationalDeviceProxy(DeviceProxyInitParams & params, PeerId peerId) : mSecureSession(*this)
     {
-        VerifyOrReturn(params.Validate() == CHIP_NO_ERROR);
+        mInitParams = params;
+        // Do not do worse
```
[the remark this answers is not in the document]
Comment (tcarmelveilleux) [reply src/app/OperationalDeviceProxy.h:98]: Fixed

File: src/app/tests/TestOperationalDeviceProxy.cpp [hunk 17]
```diff
@@ -40,6 +42,7 @@ using TestTransportMgr = TransportMgr<Transport::UDP>;
 
 void TestOperationalDeviceProxy_EstablishSessionDirectly(nlTestSuite * inSuite, void * inContext)
 {
+    // TODO: This test appears not to be workable since it does not init the fabric table!!!
```
[the remark this answers is not in the document]
Comment (tcarmelveilleux) [reply src/app/tests/TestOperationalDeviceProxy.cpp:45]: https://github.com/project-chip/connectedhomeip/issues/16845

File: src/controller/ExampleOperationalCredentialsIssuer.cpp [hunk 18]
```diff
@@ -229,9 +230,27 @@ CHIP_ERROR ExampleOperationalCredentialsIssuer::GenerateNOCChain(const ByteSpan
...
+    // Prepare IPK to be sent back. A more fully-fledged operational credentials delegate
+    // would obtain a suitable key per fabric.
+    uint8_t ipkValue[CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES];
+    Crypto::AesCcm128KeySpan ipkSpan(ipkValue);
+
+    ReturnErrorCodeIf(defaultIpkSpan.size() != sizeof(ipkValue), CHIP_ERROR_INTERNAL);
+
+    memcpy(&ipkValue[0], defaultIpkSpan.data(), defaultIpkSpan.size());
+    Optional<Crypto::AesCcm128KeySpan> ipkSpanValue;
+    ipkSpanValue.SetValue(ipkSpan);
```
[the remark this answers is not in the document]
Comment (tcarmelveilleux) [reply src/controller/ExampleOperationalCredentialsIssuer.cpp:249]: Done

File: src/credentials/FabricTable.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/credentials/FabricTable.cpp:290]: OK, sounds good.

File: src/credentials/GroupDataProviderImpl.cpp [hunk 19]
```diff
@@ -1821,6 +1823,37 @@ Crypto::SymmetricKeyContext * GroupDataProviderImpl::GetKeyContext(FabricIndex f
...
+    VerifyOrReturnError(keyset.Find(mStorage, fabric, kIdentityProtectionKeySetId), CHIP_ERROR_NOT_FOUND);
+
+    // If the keyset ID doesn't match, we have a ... problem.
+    VerifyOrReturnError(keyset.keyset_id == kIdentityProtectionKeySetId, CHIP_ERROR_INTERNAL);
+
+    out_keyset.keyset_id     = keyset.keyset_id;
+    out_keyset.num_keys_used = keyset.keys_count;
+    out_keyset.policy        = keyset.policy;
+
+    for (size_t key_idx = 0; key_idx < KeySet::kEpochKeysMax; ++key_idx)
```
[the remark this answers is not in the document]
Comment (tcarmelveilleux) [reply src/credentials/GroupDataProviderImpl.cpp:1844]: Done

File: src/protocols/secure_channel/CASEDestinationId.cpp [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+    bbuf.Put64(fabricId);
+    bbuf.Put64(nodeId);
+
+    size_t written = 0;
+    VerifyOrReturnError(bbuf.Fit(written), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    HMAC_sha hmac;
+    CHIP_ERROR err =
+        hmac.HMAC_SHA256(ipk.data(), ipk.size(), bbuf.Buffer(), written, outDestinationId.data(), outDestinationId.size());
+    return err;
```
[the remark this answers is not in the document]
Comment (tcarmelveilleux) [reply src/protocols/secure_channel/CASEDestinationId.cpp:55]: Done

File: src/protocols/secure_channel/CASEDestinationId.h [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+#include <crypto/CHIPCryptoPAL.h>
+
+#include <lib/core/CHIPError.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/support/Span.h>
+
+namespace chip {
+
+constexpr uint16_t kSigmaParamRandomNumberSize = 32;
+constexpr uint16_t kIPKSize                    = Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES;
```
[the remark this answers is not in the document]
Comment (tcarmelveilleux) [reply src/protocols/secure_channel/CASEDestinationId.h:32]: May do in a follow-up. This is scarcely used.

File: src/protocols/secure_channel/CASESession.h (hunk 14 above)
Comment (tcarmelveilleux) [reply src/protocols/secure_channel/CASESession.h:198]: Done
